### PR TITLE
add transaction isolation

### PR DESCRIPTION
**Pull request by wentaoxu**
PTAL @shenli @disksing @siddontang @coocood

### CODE REVIEW

File: sql/transaction_isolation.md [hunk 1]
```diff
@@ -0,0 +1,68 @@
+# tidb事务隔离级别
```
Comment (siddontang) [inline sql/transaction_isolation.md]: TiDB

File: sql/transaction_isolation.md [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+事务隔离级别是数据库事务处理的基础，ACID中I，即isolation，指的就是事务的隔离性。
```
Comment (siddontang) [inline sql/transaction_isolation.md]: 中英文空格

File: sql/transaction_isolation.md [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+sql 92标准定义了4种隔离级别，读未提交，读一提交，可重复读，串行化，见下表。
```
Comment (siddontang) [inline sql/transaction_isolation.md]: SQL

**Comment by wentaoxu** [comment]
PTAL @lilin90 @siddontang

File: sql/transaction-isolation.md [hunk 4]
```diff
@@ -0,0 +1,73 @@
+---
+title: TIDB 事务隔离级别
```
Comment (lilin90) [inline sql/transaction-isolation.md]: TIDB -> TiDB

File: sql/transaction-isolation.md [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+category: user guide
+---
+
+
+# TIDB 事务隔离级别
```
Comment (lilin90) [inline sql/transaction-isolation.md]: Ditto.

File: sql/transaction-isolation.md [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+
+事务隔离级别是数据库事务处理的基础，ACID 中 I，即 isolation，指的就是事务的隔离性。
+
+sql 92标准定义了4种隔离级别，读未提交，读已提交，可重复读，串行化，见下表。
```
Comment (lilin90) [inline sql/transaction-isolation.md]: Change this sentence to:
SQL 92 标准定义了 4 种隔离级别：读未提交、读已提交、可重复读、串行化。详见下表：

File: sql/transaction-isolation.md [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+
+| isolation Level  | Dirty Read   | Nonrepeatable Read | Phantom Read          | Serialization Anomaly |
```
Comment (lilin90) [inline sql/transaction-isolation.md]: isolation Level -> Isolation Level

File: sql/transaction-isolation.md [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+| ---------------- | ------------ | ------------------ | --------------------- | --------------------- |
+| Read uncommitted | Possible     | Possible           | Possible              | Possible              |
+| Read committed   | Not possible | Possible           | Possible              | Possible              |
+| Repeatable read  | Not possible | Not possible       | Not possible in  TIDB | Possible              |
+| Serializable     | Not possible | Not possible       | Not possible          | Not possible          |
+
+TIDB实现了其中的两种，读已提交和串行化。
```
Comment (lilin90) [inline sql/transaction-isolation.md]: 1. TIDB -> TiDB
2. Leave a whitespace between English and Chinese characters.
3. Please fix all TIDB -> TiDB, and add whitespace between English and Chinese characters in this PR.

File: sql/transaction-isolation.md [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+
+TIDB使用[percolator事务模型](https://research.google.com/pubs/pub36726.html)，当事务启动时会获取全局读时间戳，事务提交时或获取全局提交时间戳，并以此确定事务的执行顺序，如果想了解TIDB事务模型的实现可以详细阅读以下两篇文章[TiKV 的 MVCC（Multi-Version Concurrency Control）机制](https://pingcap.com/blog-cn/mvcc-in-tikv/)，[Percolator 和 TiDB 事务算法](https://pingcap.com/blog-cn/percolator-and-txn/)。
```
Comment (lilin90) [inline sql/transaction-isolation.md]: 或 -> 会

File: sql/transaction-isolation.md [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+
+可以通过```SET SESSION TRANSACTION ISOLATION LEVEL [read committed|repeatable read]```的命令设置事务的隔离级别。
```
Comment (lilin90) [inline sql/transaction-isolation.md]: Do not use code block style within a sentence. Change this to:

可以通过以下命令设置事务的隔离级别：

```
SET SESSION TRANSACTION ISOLATION LEVEL [read committed|repeatable read]
```

File: sql/transaction-isolation.md [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+
+
+## 可重复读
+
+可重复读是TIDB的默认隔离级别，当事务隔离级别为可重复读时，只能读到该事务启动时已经提交的其他事务修改的数据，未提交的数据或在事务启动后其他事务提交的数据是不可见的。对于本事务而言，事务语句可以看到之前的语句做出的修改。
+
+对于运行于不同节点的事务而言，不同事务启动和提交的顺序取决于从 PD 获取时间戳的顺序。
+
+处于可重复读隔离级别的事务不能并发的更新同一行，当时事务提交时发现改行在该事务启动后，已经被另一个已提交的事务更新过，那么该事务会回滚并启动自动重试。见下面这个例子。
```
Comment (lilin90) [inline sql/transaction-isolation.md]: 改行 -> 该行

File: sql/transaction-isolation.md [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+start transaction;              |               start transaction;
+select * from t1;               |               select * from t1;
+update t1 set id=id+1;          |               update t1 set id=id+1;
+commit;                         |
+                                |               commit; --回滚并自动重试
+```
+
+### 与 ANSI 可重复读隔离级别的区别
+
+尽管名称是可重复读隔离级别，但是 TIDB 中可重复读隔离级别和 ANSI 可重复隔离级别是不同的，按照[A Critique of ANSI SQL Isolation Levels](https://www.microsoft.com/en-us/research/wp-content/uploads/2016/02/tr-95-51.pdf)论文中的标准，TIDB 实现的是论文中的 snapshot 隔离级别，该隔离级别不会出现幻读，但是会出现写偏斜，而 ANSI 可重复读隔离级别不会出现写偏斜，会出现幻读
```
Comment (lilin90) [inline sql/transaction-isolation.md]: Add a period in the end.

File: sql/transaction-isolation.md [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+
+### 与MYSQL可重复读隔离级别的区别
```
Comment (lilin90) [inline sql/transaction-isolation.md]: Change all "MYSQL" in this section to "MySQL".

File: sql/transaction-isolation.md [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+
+MYSQL 可重复读隔离级别在更新时并不检验当前版本是否是可见的，也就是说，即使该行在事务启动后被更新过，同样可以继续更新。这种情况在 TIDB 会导致事务回滚并后台重试，重试最终可能会失败，导致事务最终失败，而 MYSQL 是可以更新成功的。
+MYSQL 的可重复读隔离级别并非 snapshot 隔离级别，MYSQL 可重复读隔离级别的一致性要弱于 snapshot 隔离级别，也弱于 TIDB 的可重复读隔离级别。
+
+## 读已提交
+
+读已提交隔离剂和可重复读隔离级别不同，它仅仅保证不能读到未提交事务的数据，需要注意的是，事务提交是一个动态的过程，因此读已提交隔离级别可能读到某个事务部分提交的数据。
```
Comment (lilin90) [inline sql/transaction-isolation.md]: 隔离剂 -> 隔离级别

File: sql/transaction-isolation.md [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+
+不推荐在有严格一致要求的数据库中使用读已提交隔离级别。
+
+## 事务重试
+
+对于 insert/delete/update 操作，如果事务执行失败，并且系统判断该错误为可重试，会在系统内部自动重试事务。
+
+你可以通过配置参数```retry-limit```控制自动重试的次数
```
Comment (lilin90) [inline sql/transaction-isolation.md]: Change this sentence to:

通过配置参数 `retry-limit` 可控制自动重试的次数：

File: sql/transaction-isolation.md [hunk 16]
```diff
@@ -0,0 +1,73 @@
...
+                                |               commit; --回滚并自动重试
+```
+
+### 与 ANSI 可重复读隔离级别的区别
+
+尽管名称是可重复读隔离级别，但是 TIDB 中可重复读隔离级别和 ANSI 可重复隔离级别是不同的，按照[A Critique of ANSI SQL Isolation Levels](https://www.microsoft.com/en-us/research/wp-content/uploads/2016/02/tr-95-51.pdf)论文中的标准，TIDB 实现的是论文中的 snapshot 隔离级别，该隔离级别不会出现幻读，但是会出现写偏斜，而 ANSI 可重复读隔离级别不会出现写偏斜，会出现幻读
+
+### 与MYSQL可重复读隔离级别的区别
+
+MYSQL 可重复读隔离级别在更新时并不检验当前版本是否是可见的，也就是说，即使该行在事务启动后被更新过，同样可以继续更新。这种情况在 TIDB 会导致事务回滚并后台重试，重试最终可能会失败，导致事务最终失败，而 MYSQL 是可以更新成功的。
```
Comment (lilin90) [inline sql/transaction-isolation.md]: “是否是可见的” -> “是否可见”

File: sql/transaction-isolation.md (hunk 8 above)
Comment (coocood) [inline sql/transaction-isolation.md]: 没有实现 Serializeble

File: sql/transaction-isolation.md (hunk 4 above)
Comment (wentaoxu) [reply sql/transaction-isolation.md]: ok

File: sql/transaction-isolation.md (hunk 5 above)
Comment (wentaoxu) [reply sql/transaction-isolation.md]: ok

File: sql/transaction-isolation.md (hunk 6 above)
Comment (wentaoxu) [reply sql/transaction-isolation.md]: ok

File: sql/transaction-isolation.md (hunk 7 above)
Comment (wentaoxu) [reply sql/transaction-isolation.md]: ok

File: sql/transaction-isolation.md (hunk 8 above)
Comment (wentaoxu) [reply sql/transaction-isolation.md]: ok

File: sql/transaction-isolation.md (hunk 9 above)
Comment (wentaoxu) [reply sql/transaction-isolation.md]: ok

File: sql/transaction-isolation.md (hunk 10 above)
Comment (wentaoxu) [reply sql/transaction-isolation.md]: ok

File: sql/transaction-isolation.md (hunk 11 above)
Comment (wentaoxu) [reply sql/transaction-isolation.md]: ok

File: sql/transaction-isolation.md (hunk 12 above)
Comment (wentaoxu) [reply sql/transaction-isolation.md]: ok

File: sql/transaction-isolation.md (hunk 13 above)
Comment (wentaoxu) [reply sql/transaction-isolation.md]: ok

File: sql/transaction-isolation.md (hunk 14 above)
Comment (wentaoxu) [reply sql/transaction-isolation.md]: ok

File: sql/transaction-isolation.md (hunk 15 above)
Comment (wentaoxu) [reply sql/transaction-isolation.md]: ok

File: sql/transaction-isolation.md (hunk 16 above)
Comment (wentaoxu) [reply sql/transaction-isolation.md]: ok

File: sql/transaction-isolation.md (hunk 8 above)
Comment (wentaoxu) [reply sql/transaction-isolation.md]: ok

**Comment by wentaoxu** [comment]
PTAL @coocood @lilin90

**Comment by lilin90** [comment]
LGTM

**Pull request merged**